identity-flow/design.md

```diff
@@ -0,0 +1,353 @@
+    suspend fun <NotificationT : Any> register(
+        flowName: String,
+        opaqueState: DataItem,
+        notifications: MutableSharedFlow<NotificationT>,
+        deserializer: (DataItem) -> NotificationT
+    )
+    suspend fun unregister(flowName: String, opaqueState: DataItem)
+}
+```
+
+An instance of a class implementing this must be passed to the stub implementation constructor.
+
+The important part here is that the client registers for notifications emitted to a particular
+flow instance identified by its flow name and opaque state (encrypted and serialized flow state).
+On the server, however when notification is sent it is identified by the flow name and Cbor
+representation of the state (i.e. not serialized to the sequence of bytes and not encrypted).
+
+`FlowNotificationsLocal` is an implementation for both of these interfaces which is suitable for
+the case when server-side code is being run locally.
+
+When Flow RPC operates over HTTP, a simple implementation of `FlowNotifier` can be
+be created like this `FlowNotifierPoll(FlowPollHttp(httpTransport))`. It uses long poll over a
+regular HTTP connection. On the server `FlowNotificationsLocalPoll(cipher)` creates an object
+that implements both `FlowNotifications` interface needed by `FlowEnvironment` and `FlowPoll`
+interface that needs to be passed as a parameter to `HttpHandler` constructor.
+
+Long-polling works in the following way: client collects all the listening flows that
+could be notified (flow "listens" from the time it is first returned to the client and
+until `FlowBase.complete` is called). Then the client sends a POST request using `HttpTransport`
+interface to this URL: "_/poll". The body of this request is serialized Cbor array where
+the first element is "consume token" `Tstr` (initially an empty string) and then for each
+listening flow a `Tstr` representing flow path, followed by `Bstr` that holds flow's opaque
+state. Server then responds with the serialized Cbor array containing 3 elements: updated
+"consume token" for the next long poll, zero-based index of the flow to be notified, and
+Cbor representation of the notification object. If there are no notifications, server
+holds off replying until there is one - or times-out the request. The client detects the
+time-out response and sends the poll request again using the same "consume token". Client
+also can abort the request and send a new one if a flow enters or exists listening state
+or its opaque state changes (as a result of an RPC call).
+
+Note that in high-volume environments these interfaces will need to be implemented through
+something like Android notifications.
+
+## Authentication and security issues
+
+Flow RPC does not by itself has any user authentication mechanism. However it is fairly easy
+to integrate it with an existing authentication, by adding authenticated user id as part of the
+flow state. Since flow state is encrypted and opaque to the client, it automatically serves
```

Comment:
Server state is encrypted...but do we do anything to ensure its integrity? If a client flips a few bits in the opaque blob, would we detect it as invalid state, or is it possible for the client to send an altered state back to the server?

Reply:
Added clarification that authenticated encryption is needed for this